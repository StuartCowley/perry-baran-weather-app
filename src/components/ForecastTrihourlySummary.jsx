import React from "react";
import PropTypes from "prop-types";
import WeatherIcon from "react-icons-weather";
import "../styles/ForecastTrihourlySummary.css";
import getUnits from "../helpers/getUnits";
import useUnitContext from "../hooks/useUnitContext";

function ForecastTrihourlySummary({ time, weather, icon, temp }) {
  const { units } = useUnitContext();
  const { tempUnits } = getUnits(units);
  const weatherDescription = weather ? weather.split(" ") : [weather];

  return (
    <div
      className="forecast-trihourly-summary"
      data-testid="forecast-trihourly-summary"
    >
      <p>{time}</p>
      {!icon || !weather || !temp ? (
        <p>
          No <br />
          Data
        </p>
      ) : (
        <>
          <div
            className="forecast-trihourly-summary__icon"
            data-testid="forecast-icon"
          >
            <WeatherIcon name="owm" iconId={icon} />
          </div>
          <p>
            {weatherDescription[0]} <br />
            {weatherDescription[1]}
          </p>
          <p>
            {temp.toFixed(1)}
            {tempUnits}
          </p>
        </>
      )}
    </div>
  );
}

ForecastTrihourlySummary.defaultProps = {
  icon: null,
  temp: null,
  weather: null,
};

ForecastTrihourlySummary.propTypes = {
  time: PropTypes.string.isRequired,
  icon: PropTypes.string,
  temp: PropTypes.number,
  weather: PropTypes.string,
};

export default ForecastTrihourlySummary;
