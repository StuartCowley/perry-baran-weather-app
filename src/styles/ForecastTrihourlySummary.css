.forecast-trihourly-summary {
  display: flex;
  flex-direction: column;
  justify-items: center;
  align-items: center;
  gap: 10px;
  width: 100%;
  padding-bottom: 10px;
  padding-top: 5px;
  border-left: 4px solid var(--white-blue);
  text-align: center;
}

.forecast-trihourly-summary:first-child {
  border-left: none;
}

.forecast-trihourly-summary > p {
  font-size: min(1rem, 2.8cqw);
  text-shadow: 1px 1px 2px var(--almost-white);
}

.forecast-trihourly-summary__icon {
  font-size: min(4rem, 5cqw);
}