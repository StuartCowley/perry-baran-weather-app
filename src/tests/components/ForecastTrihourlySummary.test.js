import React from "react";
import { render } from "@testing-library/react";
import ForecastTrihourlySummary from "../../components/ForecastTrihourlySummary";

describe("ForecastTrihourlySummary", () => {
  const validProps = {
    time: "15:00",
    icon: "200",
    temp: 34,
    weather: "broken clouds",
  };
  const { time, icon, temp, weather } = validProps;

  describe("all data present", () => {
    let screen;

    beforeEach(() => {
      screen = render(
        <ForecastTrihourlySummary
          time={time}
          icon={icon}
          temp={temp}
          weather={weather}
        />
      );
    });

    test("snapshot", () => {
      const { asFragment } = screen;

      expect(asFragment()).toMatchSnapshot();
    });

    test("correctly renders props", () => {
      const { getByText, getByTestId } = screen;
      const description = weather.split(" ");

      expect(getByText(time)).toBeInstanceOf(HTMLParagraphElement);
      expect(getByTestId("forecast-icon")).toContainHTML("i");
      expect(getByText(new RegExp(description[0]))).toBeInstanceOf(
        HTMLParagraphElement
      );
      expect(getByText(new RegExp(description[1]))).toBeInstanceOf(
        HTMLParagraphElement
      );
      expect(getByText(new RegExp(temp))).toBeInstanceOf(HTMLParagraphElement);
    });
  });

  describe("missing data", () => {
    test("returns No Data if icon is undefined", () => {
      const { getByText } = render(
        <ForecastTrihourlySummary time={time} temp={temp} weather={weather} />
      );

      expect(getByText(/^no.data$/i)).toBeInTheDocument();
    });

    test("returns No Data if weather is undefined", () => {
      const { getByText } = render(
        <ForecastTrihourlySummary time={time} temp={temp} icon={icon} />
      );

      expect(getByText(/^no.data$/i)).toBeInTheDocument();
    });

    test("returns No Data if temp is undefined", () => {
      const { getByText } = render(
        <ForecastTrihourlySummary time={time} icon={icon} weather={weather} />
      );

      expect(getByText(/^no.data$/i)).toBeInTheDocument();
    });
  });
});
