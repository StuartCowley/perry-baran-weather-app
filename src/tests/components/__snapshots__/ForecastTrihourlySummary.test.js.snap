// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastTrihourlySummary all data present snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-trihourly-summary"
    data-testid="forecast-trihourly-summary"
  >
    <p>
      15:00
    </p>
    <div
      class="forecast-trihourly-summary__icon"
      data-testid="forecast-icon"
    >
      <i
        class="wi wi-day-storm-showers"
        name="owm"
      />
    </div>
    <p>
      broken 
      <br />
      clouds
    </p>
    <p>
      34.0Â°C
    </p>
  </div>
</DocumentFragment>
`;
